### ISSUE TITLE
IsPrime function unnecessary, GetPrime function inefficient.

### ISSUE THREAD

**Issue by SunnyWar** [dotnet/corefx#2727]
In System.Collections.HashHelpers and System.Collections.NonGeneric\src\System\Collections\Hashtable and
System.Private.DataContractSerialization\src\System\Runtime\Serialization\ObjectToIdCache.cs

there are copies of the same function: private static bool IsPrime(int candidate)
      
It is used by: private static int GetPrime(int min)
   
Which is called by: Initialize, ExpandPrime, and TrimExcess

They all use a table in GetPrime: E.g. int[] primes = {3, 7, 17, 37, 89, 197, 431, 919, 1931,...7199369}.
Notice that each successive prime is larger than twice the preceding one. This appears to be because hashtables are doubled in size when they need to be expanded. But they are only allowed to expand up to 2G. The GetPrime function calls IsPrime to find a prime when the search goes above the limit of the primes table. The ExpandPrime function is the most interesting as it more than doubles the size each time:
public static int ExpandPrime(int oldSize)
        {
            int newSize = 2 * oldSize;
            ....
         return GetPrime(newSize);

Therefore it appears that we don't need the IsPrime function at all. Instead we simply need to expand the primes table up to 2146435069 (largest prime below 2G).  {..., 7199369, 21598111, 64794337, 194383039, 583149179, 1749447569, 953375459, 2146435069};

Recommendation:
1) Use a common GetPrime(int min) function instead of having three copies
2) Expand the primes table as described above.
3) Remove the IsPrime function, modify GetPrime to only use the table.

**Comment by SunnyWar** [comment]
In looking over the tables, I see they are not all composed of the same primes: odd. The one in HashHelpers.cs and Hashtable.cs appears to have primes that are increase by a factor of 1.2. The table in ObjectToldCache.cs is the only one that doubles in it's values. I assume there is some reason for the 1.2 factor?. Perhaps someone can explain this??

In any case, I re-calculated the primes based on this 1.2 factor which gives these as the complete set:

3, 7, 11, 17, 23, 29, 37, 47, 59, 71, 89, 107, 131, 163, 197, 239, 293, 353, 431, 521, 631, 761, 919, 1103, 1327, 1597, 1931, 2333, 2801, 3371, 4049, 4861, 5839, 7013, 8419, 10103, 12143, 14591, 17519, 21023, 25229, 30293, 36353, 43627, 52361, 62851, 75431, 90523, 108631, 130363, 156437, 187751, 225307, 270371, 324449, 389357, 467237, 560689, 672827, 807403, 968897, 1162687, 1395263, 1674319, 2009191, 2411033, 2893249, 3471899, 4166287, 4999559, 5999471, 7199369, 8639249, 10367101, 12440537, 14928671, 17914409, 21497293, 25796759, 30956117, 37147349, 44576837, 53492207, 64190669, 77028803, 92434613, 110921543, 133105859, 159727031, 191672443, 230006941, 276008387, 331210079, 397452101, 476942527, 572331049, 686797261, 824156741, 988988137, 1186785773, 1424142949, 1708971541, 2050765853, 2146435069

**Comment by weshaggard** [comment]
@vancem @ellismg either of you guys know the reasoning behind the prime table?

**Comment by vancem** [comment]
As far as 1.2 vs 2.0, that is all about the tradeoff of space for time (you waste more time rehashing with 1.2 but you waste less space).   It should probably be close to 2.0 for small numbers but then convert to 1.2 as you get bigger.    

Eliminating the dynamic calculation of primes is a nice simplification.

**Comment by SunnyWar** [comment]
@vancem I agree. Is there any data floating around on how large hash tables typically get? I believe the difference in resizing times is ~log(eventualSize)/log(factor).. E.g. with and eventual size of 10,000 the factor of 1.2 would resize near to 50 times and a factor of 2.0 would do it only 13 times. I think the actual numbers are 35 and 11 but I don't know how to account for finding the next nearest prime. The formula is good enough. IMHO.

Here's first bunch of primes below 100,000 in case we want to use some of them to make the lower hashtable resizes more efficient.

3, 7, 17, 37, 79, 163, 331, 673, 1361, 2729, 5471, 10949, 21911, 43853, 87719, 175447, 350899, 701819

**Comment by SunnyWar** [comment]
I have a code change for this ready to go. I just have to figure out how to make GitHub work.

**Comment by hackcraft** [comment]
@SunnyWar are you going to submit a PR?

I wonder how the linear search this contains:

    for (int i = 0; i < primes.Length; i++)
    {
        int prime = primes[i];
        if (prime >= min) return prime;
    }

Compares with a binary search:

    int lo = 0;
    int hi = primes.Length - 1;
    while (lo <= hi)
    {
        int i = lo + ((hi - lo) >> 1);
        int item = primes[i];
    
        if (item == min) return i;
        if (item < min) lo = i + 1;
        else hi = i - 1;
    }
    return lo != primes.Length ? primes[lo] : min;

**Comment by redknightlois** [comment]
@hackcraft I was going to look into that particular case too because we had a particular scenario where that is causing a huge slowdown (specially with "small" increments). Binary search will probably work worse for the normal case (you create a dictionary and start adding data into it). 

The PR I had in mind but never got the time to do is to use an indexing table for the first 512 items. So the code would look like this:

```
if ( size < 512 )
    return smallPrimesTable[size];
else
    return GetPrimes();
```

We remove the first 512 primers from the original table. And then only hit .GetPrimes() when the size of the resize is big (where GetPrimes() cost will be depreciable).

The smallPrimesTable is just an unrolled version of the original table that will take 2Kb of memory in exchange of a direct access to the correct prime used.

```readonly int[] smallPrimesTable = new[512] {  3, 3, 3, 3, 7, 7, 7, 7, 11, 11, 11, 11, ...... ```

And so on.

**Comment by hackcraft** [comment]
@redknightlois I experimented with binary search since asking that comment above, and it's not a great idea; you have to be high into the list before it breaks even with linear search, and very far into it before it's an improvement, at which point it's small potatoes as the resize operation as a whole goes.

Did you do measurements on your lookup idea?

**Comment by redknightlois** [comment]
@hackcraft No, I didnt experiment with it. But profiling runs on the scenario where I uncovered this suggest that for the first 512 it is a net win. You are trading a single indirect load operation for a "for loop" (several more operations). You will have a potential cache miss in both situations so you are on par on that side too.

**Comment by mikedn** [comment]
It's very unlikely that optimizing `GetPrime` like would provide any measureable benefit. Remember that `GetPrime` is called when the hashtable capacity is increased and as with all collections the capacity doesn't increase often. More importantly, increasing the capacity of a hashtable is a rather expensive operation, much more expensive than that trivial linear search.

**Comment by hackcraft** [comment]
@mikedn The idea that started this issue certainly does when the slower path is hit, as the approach taken is quite expensive. Sure, its about to do a large copy after that, but calculating primes is definitely busywork prior to that; enough that it's sometimes used as a keep-the-cpu-busy task in tests.

I'd be more suspicious of @redknightlois' idea, but they say they've found it does help in practice.

**Comment by redknightlois** [comment]
@mikedn We actually implemented a replacement for that specific scenario with an entirely different dictionary algorithm that is slower for most uses, but faster for that specific workload. If the GetPrimes() function is fixed there is no need for us to build an entirely new dictionary collection because we can easily use the Dictionary( Size ) constructor without the GetPrimes() impact.  

We are pretty certain that we have hit an uncommon scenario, but these are examples on that particular workflow (the one that has to perform multiple resizes from very small sizes) vs the one that we implemented to avoid that. 

![fast-dictionary](https://cloud.githubusercontent.com/assets/1045411/9446578/d1281fca-4a68-11e5-8b1c-9c05433994b4.png)

![fast-dictionary-2](https://cloud.githubusercontent.com/assets/1045411/9446579/d1f170b4-4a68-11e5-9f08-f56b4e992bce.png)

Moreover, the change has this impact in the whole system:

![fast-dictionary-impact](https://cloud.githubusercontent.com/assets/1045411/9446736/948b154e-4a69-11e5-85bb-341d2b41c005.png)

**Comment by mikedn** [comment]
@JonHanna I was mostly referring to @redknightlois proposal myself. As for @SunnyWar proposal I think I'd choose to do it simply because it simplifies the code. Performance - maybe but keep in mind that the slow path is taken for very large hashtables. If you have to deal with such large hashtables then you'll likely have more serious issues due to LOH usage. And no, hashtable resizing is not any large array copy, copying the buckets array involves computing indices via expensive modulo operations.

**Comment by mikedn** [comment]
@redknightlois 

> If the GetPrimes() function is fixed there is no need for us to build an entirely new dictionary collection because we can easily use the Dictionary( Size ) constructor without the GetPrimes() impact.

I see, you are saying that the main problem in your case that you want to build hashtables with a specific capacity and the prime number lookup is significant in this case. That sounds reasonable but I don't know what to make of your profile traces, none of them appear to be related to GetPrimes.

**Comment by redknightlois** [comment]
@mikedn Believe me that noone takes the path of implementing a dictionary collection if there is no evidence the problem is there :) ... If we try to be clever and use the Dictionary(Size) constructor we are actually losing because of GetPrimes(). The issue is with small size dictionaries where the allocation time is despreciable against the GetPrimes() cost. In the general case of big dictionaries, you wont see any statistical difference. 

Because they are not shown, Insert will cause the resize and therefore the GetPrimes() call. In the alternative case we are using power-of-two (and therefore a different algorithm to select the bucket) but finding the size in that case is very fast. 

I mistakenly added also the TryGet() which is not really important for this. But that showcases how a very small improvement can have a huge difference in time. In this case, the inner call to Insert not being inlined in the set_item indexer.

**Comment by mikedn** [comment]
@redknightlois 

> In the alternative case we are using power-of-two (and therefore a different algorithm to select the bucket) but finding the size in that case is very fast. 

Wait a moment, are you saying that you made the size a power of two and thus got rid of the modulo operation??

**Comment by redknightlois** [comment]
@mikedn Yes, but for that we had to change the underlying algorithm. That is why it is worse for most cases except our particular workload. Dictionary<T> modulo based algorithm is still faster for most of the use cases.

**Comment by JonHanna** [comment]
> Believe me that noone takes the path of implementing a dictionary collection if there is no evidence the problem is there

I don't know, the sort of person who submits pull requests to this repository might. On which note,  power-of-two vs prime has some complicated tradeoffs that throw up surprises. 

On less complicated and more relevant matters, are we going to see a pull request on this?

**Comment by redknightlois** [comment]
@JonHanna Yes, based on our work on that regard, I dont think it would make sense to change the Dictionary<T> implementation would make sense (except for a few tweaks that can be made like the GetPrimes()) because it is pretty good overall.

**Comment by mikedn** [comment]
@redknightlois The "problem" with that you did is that you replaced the most expensive general purpose CPU instruction with a much cheaper one. Because of that there's a good chance that your implementation will be faster even if GetPrime is improved.

Anyway, 1 million GetPrime(100) take ~76ms (Haswell Core i5 3.1GHz) with the original code and ~21ms with @redknightlois's suggestion. For those who create many dictionaries and specify a small initial capacity that would be an improvement.

And 1 mil GetPrime(7000000) take ~400ms while 1 mil GetPrime(8000000) take ~6500ms so completing the rest of the table as @SunnyWar suggested would be an improvement, at least when taken in isolation.

**Comment by redknightlois** [comment]
@mikedn hitting 1 million is not that difficult. Any resize will issue a call to GetPrime(x), then if you create lots of small dictionaries and insert (say 13 elements in average) each resize will do a call there (that's 2 or 3 resizes if I recall properly).  

About the power of two, actually no, the stock .Net version is faster for almost every scenario out there except the create many small dictionaries.

PS: I couldnt make it faster, if for the fun of it you want to give it a try here is the source code:  https://github.com/Corvalius/ravendb/blob/c56aae2da6c7b374759c462281d704dca1af64a4/Raven.Abstractions/Util/FastDictionary.cs

**Comment by mikedn** [comment]
@redknightlois Well, it's well known that using power of two for hashtables but the point is that it works well in your scenario and you took advantage of that and wrote a completely different hashtable. And if GetPrime is improved there's a very good chance that your implementation is still faster.

**Comment by JonHanna** [comment]
Off-topic, but if you're going to use power of two, it can be well to redistribute the hash code before the (mask-based) modulo, as poor hash distribution hurts more, but many GetHashCode implementations have good overall distribution, but poor distribution within just the lower bits.

**Comment by redknightlois** [comment]
@JonHanna Yes, took care of that using xxHash when calculating the hash ;)

**Comment by JonHanna** [comment]
I was led to write the "SpookilySharp" port of SpookyHash by having a reason for having a power-of-two linear reprobing hashtable, and being absolutely hammered by bad hash methods.

**Comment by JonHanna** [comment]
Coming back to the original issue here, I don't think the suggested change is a good idea.

The values suggested are just wrong for `ObjectToIdCache`, but I'll come back to that as I think there is a point in changing its `GetPrime()` but in a different way.

Trying the suggested values just running the tests in /src/System.Collections takes considerably longer.

Changing the values to precomputing those that would be arrived at by the way the collections double when growing, the tests take about the same time as before.

Such values therefore **could** turn out to be a win with more detailed profiling.

However, the sizes used if one either starts a collection larger than the largest currently pre-computed prime, or if one calls `HashSet<T>.TrimExcess()` when above such a size are now larger. Such cases would suffer from the proposed change. To justify this, speed improvements would have to not only be demonstrated but be considerable.

As such the proposed change would be a nett loss.

`ObjectToIdCache` on the other hand always starts with a size of `3` and then doubles each time. We can hence pre-calculate the values it would currently use, and also increase its robustness in the case of (unlikely) massive sizes being used. I've submitted #2978 to consider that case.

**Comment by Tydude4Christ** [comment]
I propose an implementation of GetPrime that has a prime list that is logarithmically increasing by a factor that is slightly greater than the cube root of 2 or fourth root of 2. This will allow efficient retrieval of a prime number that is just greater than double that of a number already on the prime list by simply using the log function and arithmetic. I've implemented this architecture in a library of mine at https://github.com/Tydude4Christ/Enums.NET/blob/master/Src/Enums.NET/HashHelper.cs

**Comment by SunnyWar** [comment]
Off and on over several weeks I tried many variations of what has been suggested. First I tried using a table of power of two with the index being the base2 log of the min value. I didn't use Math.Log as it's too slow but instead did some bit twiddling. I would have loved to use the C++ intrinsic _BitScanReverse but, unfortunately, such low-level optimized code is not possible (I think. PLEASE show me I'm wrong). This approach is much faster than what is currently used BUT the table size is much smaller. This means the hash bucket size would often be very far off the optimal.

So then I spent some time trying to have a large table with very good coverage in the lower bounds and a fast lookup. I make a table of primes close to 2^1.2 and used log base 1.2 of min as the index. Again, since Math.Log is slow I used the LOGb(a) = LOGd(a)/LOGd(b) and LOG(a*b) = LOG(a) + LOG(b) identities and LOG2 bit twiddling to make it faster. I was able to get about a 3X improvement in lookup speed.

I'm going to do some more testing of this approach. So far it looks promising. The drawback is that the code is much harder to understand. It's probably worth the added complexity as it's used so often.

Finally I'd like to offer that I'd prefer that when the constructor is called with a specific size that the nearest possible prime be used, rather than a number from the table, as in this scenario the developer knows the exact size needed. We should optimize for this and eat the slight cost of calculating the next nearest prime.

**Comment by vancem** [comment]
I think we are trying a bit to hard here.     In my opinion the straw man to beat is the case where you put all the primes in the table (there are 104 of them), and do a linear search (if you wish, you can put on check that does the first step of a binary search (thus if x > 968897 startIdx = 50) to half the time for large numbers.  (And only because this code does not add interesting complexity). 

But the real point here is that I don't see a actual performance issue here.     While you can imagine microbenchmarks that construct a LARGE NUMBER of  large hash tables in which this loop this can be interesting, that is not an interesting case (because the initialized it with a large number, you expect to have a large number of inserts, and compared to the insert cost the cost of the construction is small.

This is also true for the resize case.   The cost of searching the table is log(N) but the cost of actually doing the resize is N, so reasonable size N the cost of this search will NEVER be a dominate overall cost.   (Also keep in mind that in addition to be log(N) the search is also cheaper than the corresponding insert).  

Thus the right tradeoff is to do get what you can while KEEPING IT SIMPLE.   Thus a uniform table and linear search is the simple thing to do (less code, easy to understand).   This is the strawman.  

I would argue that any other design should have to justify its performance improvement with a real benchmark with real time.     The only numbers I see above are from redknightlois, the stacks don't show the construction/prime number generation at all (which is what we would be changing).  

My argument above is that you won't be able to.   Feel free to prove me wrong, but the easier thing to do is to simply fix the code based on a SIMPLICITY argument.  

Vance Morrison
.NET Runtime Performance Architect.

P.S:  one other possibility that I would entertain is to get out of the business of needing a prime at all.   It is not clear it has value.   (and again the argument here is SIMPLICITY).

**Comment by SunnyWar** [comment]
I'm at 100% for keeping the code simple in 99% of the cases. The 1% is when a body of code is called a LOT, then I'm willing to suffer readability for performance. Having said that, I'm hard pressed to name a performance essential body of code I've written in recent memory that didn't make significant use of a HashSet or a Dictionary. My anecdotal experience is that nearly every operation on all generic collections have significant performance implications. (not to mention Linq operations!)

For example: Recently I worked on server code that has a Dictionary / user on a for 10 of thousands of users. Each Dictionary contains thousands of properties about the user.  New users are coming and going every second. Properties are changed on all users several times a second.  A very small performance change in HashSet create and population is measurable. It can make a difference in the number of servers that are needed, in scalability, and ultimately the cost.

So again, in general, I agree with you. My rule is "don't be clever" but I believe in breaking the rule when it costs a lot.  In this case, hash tables are so fundamental they deserve as much optimization as we can muster.

**Comment by vancem** [comment]
@SunnyWar   My argument is NOT that Dictionaries are unimportant, it is that this code path through dictionary is unimportant.     Note that we don't need to argue about this.  If what you say is true and it is important, it should be easy for you to show me a example where this code path adds up to even 1% of some microbenchmark that models a scenario that 'looks like a plausible scenario'.    I am arguing that this will not be possible because the cost here is only paid when a Dictionary is Created or Resized, and in any real scenario the cost of INSERT AND FETCH will always dominate (since they should be O(N) (and a relatively large constant), where this cost is O(log(N)) and has a small proportionality constant.  

Notice that in absolute value yes, maybe non-trivial time can be spent in this routine, but only when 100 to 1000 time more time is being spent insertions and fetches.  

Thus I am arguing that this does not 'cost a lot' and thus simplicity should rule (we should concentrate our effort in making insertion and fetch .1% faster).  

For what it is worth...

**Comment by redknightlois** [comment]
Agree with @vancem wholeheartly. We use extensively hash based structures of basically all the classes out there even SortedDictionary (to be true we built a different structure for strings dictionaries so we are not using stock, but that is another story). 

The only class of problems where we have evidence that improving the constructor is when you are building an insane amount of small dictionaries with a known size. (see my comments from a few months ago, they even have profiling data from a real-world project). The solution for that one is just avoid GetPrime altogether for sizes less than 512 or so (even 256 would work).  

For big dictionaries, just calculating the hashes of the objects themselves would easily offset the GetPrime cost.

**Comment by SunnyWar** [comment]
@vancem ....ok. Following along with your reasoning, why have the table at all? Wouldn't it be better to have a GetNextPrime(int min) function that actually finds a prime >= min?  The cost of such a function would be quite low and primes as quite "dense". This would also make the buckets optimally sized once the table is full.

**Comment by redknightlois** [comment]
@SunnyWar Following on GetNextPrime. Why would you use a non O(1) algorithm to get it, if you can use a table and get it with a few (probably 3 assembler instructions and 1 memory read)? The total cost of ownership of such algorithm for the sizes that matter is n*4 bytes. Where n = 512 that would be just 2K of memory.

**Comment by redknightlois** [comment]
@vancem Sorry didnt realize you said this.

> I am arguing that this will not be possible because the cost here is only paid when a Dictionary is Created or Resized, and in any real scenario the cost of INSERT AND FETCH will always dominate

I have enough data that suggest this is not true for ANY condition. At least one condition where this is not true is when the dictionary is small and will not grow. In those cases the dictionary allocation, creation and resizing do dominate; when that is multiplied for million short lived dictionaries you do see the difference (as @mikedn showed before when trying out the optimization I suggested). 

BTW pooling those dictionaries, at least in the case that we investigated, was actually costlier than the allocation itself. I presume that the cost of clearing the hash table dominates against the GC doing its work.

**Comment by vancem** [comment]
@redknightlois   Indeed if you ONLY construct Dictionaries on a hot path, then the cost can be noticeable (even dominate).   However dictionaries are meant to have elements inserted into them and removed from them.   You also expect that this happens more than once.   In fact you kind of expect it to be AT LEAST O(N) (at a minimum you expect to insert enough times to fill 'most' of the dictionary's initial size (otherwise you have made your guess too big).   On top of that, you expect to fetch at least a good fraction of what you put in the Dictionary at least once (and typically much more than once).    Under these assumptions (which I would argue are needed for the scenario to be 'real world' the inseration and fetch costs will dominate.   

You will note that I did not say it is impossible for the cost of prime lookup to dominate, only that it does not seem 'real world' and thus not worth sacrificing simplicity.

@SunnyWar - it is true that by my reasoning removing the table and computing the prime instead is a reasonable solution.    You definitely could argue that.   However I take a pragmatic view that if I have two solutions that are EQUALLY simple, I will take the more efficient one even if I did not have to.   Since a lookup table is pretty simple (arguably simpler than the logic to find primes), I am OK with that solution.

**Comment by jasonwilliams200OK** [comment]
Was this fixed by #6203?

**Comment by SunnyWar** [comment]
I believe this conversation went into the weeds which is why I submitted #6203 seperately. I think this should be closed.

**Comment by jasonwilliams200OK** [comment]
Adding a separate (e.g. `"Fixes #1234."`) line in PR description or commit log auto-closes the issue on merging the PR (preferably it should be the *punchline*, and never the first line). Also works for multiple issues at once. Find more tricks and tips at https://help.github.com/articles/closing-issues-via-commit-messages/. :trident:

**Comment by SunnyWar** [comment]
thank you for the information
